### PR TITLE
[s2s trainer] tests to use distributed on multi-gpu machine

### LINKED ISSUE #7833: [s2s trainer] tests fail on multi-gpu machine
#### Command
```bash
RUN_SLOW=1 USE_CUDA=1 pytest examples/seq2seq/test_finetune_trainer.py
```

#### Traceback
```python
=========================================================== test session starts ===========================================================
platform linux -- Python 3.7.4, pytest-5.3.5, py-1.8.1, pluggy-0.13.1
rootdir: /home/shleifer/transformers_fork, inifile: pytest.ini
plugins: forked-1.1.3, hydra-core-1.0.0, xdist-1.31.0, requests-mock-1.8.0
collected 2 items

examples/seq2seq/test_finetune_trainer.py /home/shleifer/transformers_fork/src/transformers/training_args.py:339: FutureWarning: The `evaluate_during_training` argument is deprecated in favor of `evaluation_strategy` (which has more options)
  FutureWarning,
F/home/shleifer/transformers_fork/src/transformers/training_args.py:339: FutureWarning: The `evaluate_during_training` argument is deprecated in favor of `evaluation_strategy` (which has more options)
  FutureWarning,
F

================================================================ FAILURES =================================================================
__________________________________________________________ test_finetune_trainer __________________________________________________________

    def test_finetune_trainer():
>       output_dir = run_trainer(1, "12", MBART_TINY, 1)

examples/seq2seq/test_finetune_trainer.py:19:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
examples/seq2seq/test_finetune_trainer.py:105: in run_trainer
    main()
examples/seq2seq/finetune_trainer.py:294: in main
    model_path=model_args.model_name_or_path if os.path.isdir(model_args.model_name_or_path) else None
src/transformers/trainer.py:583: in train
    train_dataloader = self.get_train_dataloader()
src/transformers/trainer.py:386: in get_train_dataloader
    train_sampler = self._get_train_sampler()
examples/seq2seq/seq2seq_trainer.py:108: in _get_train_sampler
    self.args.per_device_train_batch_size, distributed=self.args.n_gpu > 1
examples/seq2seq/utils.py:156: in make_sortish_sampler
    return DistributedSortishSampler(self, batch_size, shuffle=shuffle, **kwargs)
examples/seq2seq/utils.py:368: in __init__
    num_replicas = dist.get_world_size()
../miniconda3/envs/nb/lib/python3.7/site-packages/torch/distributed/distributed_c10d.py:582: in get_world_size
    return _get_group_size(group)
../miniconda3/envs/nb/lib/python3.7/site-packages/torch/distributed/distributed_c10d.py:196: in _get_group_size
    _check_default_pg()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def _check_default_pg():
        """
        Helper that checks if the default ProcessGroup has been initialized, with
        assertion

        """
        assert _default_pg is not None, \
>           "Default process group is not initialized"
E       AssertionError: Default process group is not initialized

../miniconda3/envs/nb/lib/python3.7/site-packages/torch/distributed/distributed_c10d.py:187: AssertionError
_______________________________________________________ test_finetune_trainer_slow ________________________________________________________

    @slow
    def test_finetune_trainer_slow():
        # TODO(SS): This will fail on devices with more than 1 GPU.
        # There is a missing call to __init__process_group somewhere
>       output_dir = run_trainer(eval_steps=2, max_len="128", model_name=MARIAN_MODEL, num_train_epochs=3)

examples/seq2seq/test_finetune_trainer.py:30:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
examples/seq2seq/test_finetune_trainer.py:105: in run_trainer
    main()
examples/seq2seq/finetune_trainer.py:294: in main
    model_path=model_args.model_name_or_path if os.path.isdir(model_args.model_name_or_path) else None
src/transformers/trainer.py:583: in train
    train_dataloader = self.get_train_dataloader()
src/transformers/trainer.py:386: in get_train_dataloader
    train_sampler = self._get_train_sampler()
examples/seq2seq/seq2seq_trainer.py:108: in _get_train_sampler
    self.args.per_device_train_batch_size, distributed=self.args.n_gpu > 1
examples/seq2seq/utils.py:156: in make_sortish_sampler
    return DistributedSortishSampler(self, batch_size, shuffle=shuffle, **kwargs)
examples/seq2seq/utils.py:368: in __init__
    num_replicas = dist.get_world_size()
../miniconda3/envs/nb/lib/python3.7/site-packages/torch/distributed/distributed_c10d.py:582: in get_world_size
    return _get_group_size(group)
../miniconda3/envs/nb/lib/python3.7/site-packages/torch/distributed/distributed_c10d.py:196: in _get_group_size
    _check_default_pg()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def _check_default_pg():
        """
        Helper that checks if the default ProcessGroup has been initialized, with
        assertion

        """
        assert _default_pg is not None, \
>           "Default process group is not initialized"
E       AssertionError: Default process group is not initialized

../miniconda3/envs/nb/lib/python3.7/site-packages/torch/distributed/distributed_c10d.py:187: AssertionError
========================================================= short test summary info =========================================================
FAILED examples/seq2seq/test_finetune_trainer.py::test_finetune_trainer - AssertionError: Default process group is not initialized
FAILED examples/seq2seq/test_finetune_trainer.py::test_finetune_trainer_slow - AssertionError: Default process group is not initialized
=========================================================== 2 failed in 11.51s ============================================================
```

### LINKED ISSUE #7982: [s2s test] examples/seq2seq/test_finetune_trainer.py::TestFinetuneTrainer::test_finetune_trainer_slow fails on GPU
This works (cpu / any pytorch):
```
CUDA_VISIBLE_DEVICES="" RUN_SLOW=1 pytest -sv examples/seq2seq/test_finetune_trainer.py::TestFinetuneTrainer::test_finetune_trainer_slow
```
This fails torch-1.5/gpu or 1.6, or nightly:
```
CUDA_VISIBLE_DEVICES="0" RUN_SLOW=1 pytest -sv examples/seq2seq/test_finetune_trainer.py::TestFinetuneTrainer::test_finetune_trainer_slow
```
Same with pytorch-nightly. same with py37 and py38.

Error:
```
{'eval_loss': 5223.2333984375, 'eval_bleu': 0.0, 'eval_gen_len': 1.0, 'epoch': 1.0}                                                                   
{'eval_loss': 5064.154296875, 'eval_bleu': 0.0, 'eval_gen_len': 1.0, 'epoch': 2.0}                                                                    
{'eval_loss': 4966.837890625, 'eval_bleu': 0.0, 'eval_gen_len': 3.8, 'epoch': 3.0}                                                                    
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6 [00:03<00:00,  1.55it/s]
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  4.69it/s]
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:01<00:00,  3.04it/s]FAILED                                                                                                                                                

====================================================================== FAILURES ======================================================================
___________________________________________________ TestFinetuneTrainer.test_finetune_trainer_slow ___________________________________________________

self = <seq2seq.test_finetune_trainer.TestFinetuneTrainer testMethod=test_finetune_trainer_slow>

100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:01<00:00,  2.78it/s]
    @slow
    def test_finetune_trainer_slow(self):
        # There is a missing call to __init__process_group somewhere
        output_dir = self.run_trainer(eval_steps=2, max_len="128", model_name=MARIAN_MODEL, num_train_epochs=3)
    
        # Check metrics
        logs = TrainerState.load_from_json(os.path.join(output_dir, "trainer_state.json")).log_history
        eval_metrics = [log for log in logs if "eval_loss" in log.keys()]
        first_step_stats = eval_metrics[0]
        last_step_stats = eval_metrics[-1]
    
>       assert first_step_stats["eval_bleu"] < last_step_stats["eval_bleu"]  # model learned nothing
E       AssertionError: assert 0.0 < 0.0

examples/seq2seq/test_finetune_trainer.py:36: AssertionError
```
env:
```
- `transformers` version: 3.4.0
- Platform: Linux-4.15.0-118-generic-x86_64-with-debian-buster-sid
- Python version: 3.7.9
- PyTorch version (GPU?): 1.5.0 (True)
- Tensorflow version (GPU?): 2.3.1 (True)
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>
```

@sshleifer

### PR DESCRIPTION
**Pull request by stas00**
This PR:

* [x] abstracts the async forking io into local `utils.py`
* [x] deploys distributed training with special async io forking for `examples/seq2seq/test_finetune_trainer.py`

So now this works (2 gpus):
```
CUDA_VISIBLE_DEVICES="0,1" RUN_SLOW=1 pytest -sv examples/seq2seq/test_finetune_trainer.py
```
and this still works (1 gpu)
```
CUDA_VISIBLE_DEVICES="0" RUN_SLOW=1 pytest -sv examples/seq2seq/test_finetune_trainer.py
```

Fixes: #7833
Fixes: #7982

@sshleifer

### CODE REVIEW

**Comment by stas00** [comment]
@sshleifer, the slow test isn't working for me prior to this change:

```
CUDA_VISIBLE_DEVICES="" RUN_SLOW=1 pyt -sv examples/seq2seq/test_finetune_trainer.py::TestFinetuneTrainer::test_finetune_trainer_slow   
[...]

self = <seq2seq.test_finetune_trainer.TestFinetuneTrainer testMethod=test_finetune_trainer_slow>

    @slow
    def test_finetune_trainer_slow(self):
        # There is a missing call to __init__process_group somewhere
>       output_dir = self.run_trainer(eval_steps=2, max_len="128", model_name=MARIAN_MODEL, num_train_epochs=3)

examples/seq2seq/test_finetune_trainer.py:35: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
examples/seq2seq/test_finetune_trainer.py:113: in run_trainer
    main()
examples/seq2seq/finetune_trainer.py:199: in main
    model = AutoModelForSeq2SeqLM.from_pretrained(
src/transformers/modeling_auto.py:1118: in from_pretrained
    return MODEL_FOR_SEQ_TO_SEQ_CAUSAL_LM_MAPPING[type(config)].from_pretrained(
src/transformers/modeling_utils.py:947: in from_pretrained
    model = cls(config, *model_args, **model_kwargs)
src/transformers/modeling_bart.py:964: in __init__
    base_model = BartModel(config)
src/transformers/modeling_bart.py:843: in __init__
    self.encoder = BartEncoder(config, self.shared)
src/transformers/modeling_bart.py:315: in __init__
    self.embed_positions = SinusoidalPositionalEmbedding(
src/transformers/modeling_bart.py:1331: in __init__
    self.weight = self._init_weight(self.weight)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

out = Parameter containing:
tensor([[ 2.2244e+00, -1.2380e+00, -3.5307e-01,  ..., -1.0924e+00,
         -1.3130e+00,  1.7737...    [-3.8906e-01,  9.2203e-01,  1.7887e-01,  ..., -1.7493e-01,
         -1.6993e+00,  2.0896e-01]], requires_grad=True)

    @staticmethod
    def _init_weight(out: nn.Parameter):
        """Identical to the XLM create_sinusoidal_embeddings except features are not interleaved.
        The cos features are in the 2nd half of the vector. [dim // 2:]
        """
        n_pos, dim = out.shape
        position_enc = np.array(
            [[pos / np.power(10000, 2 * (j // 2) / dim) for j in range(dim)] for pos in range(n_pos)]
        )
>       out[:, 0 : dim // 2] = torch.FloatTensor(np.sin(position_enc[:, 0::2]))  # This line breaks for odd n_pos
E       RuntimeError: a view of a leaf Variable that requires grad is being used in an in-place operation
```

**Comment by sshleifer** [comment]
interesting, I can't reproduce that. What's your `transformers-cli env`? Does it fail after the change?

**Comment by stas00** [comment]
No, it fails on master.

```
- `transformers` version: 3.4.0
- Platform: Linux-4.15.0-118-generic-x86_64-with-glibc2.10
- Python version: 3.8.5
- PyTorch version (GPU?): 1.8.0.dev20201020 (True)
- Tensorflow version (GPU?): 2.3.1 (True)
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>
```

**Comment by sshleifer** [comment]
I'd open an issue "sinusoidal positional embedding broken on torch 1.8".
Reasoing:  [these](https://github.com/huggingface/transformers/blob/master/tests/test_modeling_bart.py#L606) fast tests pass CI and I can't replicate on on torch 1.5.

Is this ready to merge otherwise?

File: src/transformers/testing_utils.py [hunk 1]
```diff
@@ -533,3 +533,71 @@ def test_something():
...
     return unittest.mock.patch.dict(os.environ, kwargs)
+
+
+# the following code deals with async io between processes
+
+# adapted from https://stackoverflow.com/a/59041913/9201239
+import asyncio  # noqa
+
+
+class RunOutput:
```
Comment (sshleifer) [inline src/transformers/testing_utils.py:544]: maybe `ProcessOutputCollector` if it's in src/ it needs to be named more specifically I think.

Comment (stas00) [reply src/transformers/testing_utils.py:544]: it's not used outside of the wrapper - a private/internal class

Comment (stas00) [reply src/transformers/testing_utils.py:544]: as test_utils is growing this may end up in its own module

**Comment by stas00** [comment]
Need to sort out this first: https://github.com/huggingface/transformers/issues/7982

It's mostly ready otherwise, but the slow test will fail as it has nothing to do with this PR.

**Comment by sshleifer** [comment]
Out of curiosity, how did you resolve the bleu issue?

**Comment by sshleifer** [comment]
Great work, btw! This is awesome. Now we can tell people to run these tests before they break things :) 
Apparently there is multi-gpu ci running src/ tests at some frequency FYI, I think through gh actions.

**Comment by stas00** [comment]
I replied in the other issue: I used more iterations - 6 was enough for 1 gpu, 10 for 2, so I went with 10.

**Comment by stas00** [comment]
Documenting it now https://github.com/huggingface/transformers/pull/7993
If you think anything needs to be added please let me know.
It will get better over time.

**Comment by stas00** [comment]
> Apparently there is multi-gpu ci running src/ tests at some frequency FYI, I think through gh actions.
Once a day yes:
https://github.com/huggingface/transformers/blob/master/.github/workflows/self-scheduled.yml#L74

**Comment by stas00** [comment]
Later I want to add the non-interactive IO pipe options - in case it hangs for someone - by default it could be non-interactive - always works, and only make it interactive for debug purposes.

**Pull request merged**